cirq-core/cirq/ops/classically_controlled_operation.py

```diff
@@ -148,39 +163,77 @@ def _circuit_diagram_info_(
-
-        measurements = [
-            args.log_of_measurement_results.get(str(key), str(key)) for key in self._control_keys
-        ]
-        missing = [m for m in measurements if isinstance(m, str)]
-        if missing:
-            raise ValueError(f'Measurement keys {missing} missing when performing {self}')
-        if all(not_zero(measurement) for measurement in measurements):
-            protocols.act_on(self._sub_operation, args)
+        for condition in self._conditions:
+            keys, expr = condition.keys, condition.expr
+            missing = [str(k) for k in keys if str(k) not in args.log_of_measurement_results]
+            if missing:
+                raise ValueError(f'Measurement keys {missing} missing when performing {self}')
+            replacements = {
+                f'x{i}': args.log_of_measurement_results[str(k)][0] for i, k in enumerate(keys)
+            }
+            result = expr.subs(replacements)
+            if not result:
+                return True
+        protocols.act_on(self._sub_operation, args)
         return True
 
     def _with_measurement_key_mapping_(
         self, key_map: Dict[str, str]
     ) -> 'ClassicallyControlledOperation':
-        keys = [protocols.with_measurement_key_mapping(k, key_map) for k in self._control_keys]
-        return self._sub_operation.with_classical_controls(*keys)
+        def map_condition(condition: raw_types.Condition) -> raw_types.Condition:
+            keys = [protocols.with_measurement_key_mapping(k, key_map) for k in condition.keys]
+            return condition.with_keys(tuple(keys))
+        conditions = [map_condition(c) for c in self._conditions]
+        return self._sub_operation.with_classical_controls(*conditions)
 
     def _with_key_path_prefix_(self, path: Tuple[str, ...]) -> 'ClassicallyControlledOperation':
-        keys = [protocols.with_key_path_prefix(k, path) for k in self._control_keys]
-        return self._sub_operation.with_classical_controls(*keys)
+        def map_condition(condition: raw_types.Condition) -> raw_types.Condition:
+            keys = tuple(protocols.with_key_path_prefix(k, path) for k in condition.keys)
+            return condition.with_keys(keys)
+        conditions = [map_condition(c) for c in self._conditions]
+        return self._sub_operation.with_classical_controls(*conditions)
 
     def _control_keys_(self) -> FrozenSet[value.MeasurementKey]:
-        return frozenset(self._control_keys).union(protocols.control_keys(self._sub_operation))
+        local_keys = frozenset()
+        for condition in self._conditions:
+            local_keys = local_keys.union(condition.keys)
+        return local_keys.union(protocols.control_keys(self._sub_operation))
 
     def _qasm_(self, args: 'cirq.QasmArgs') -> Optional[str]:
         args.validate_version('2.0')
-        keys = [f'm_{key}!=0' for key in self._control_keys]
+        keys = [f'm_{key}!=0' for key in self._conditions]
```

Comment:
This seems broken for non-trivial expressions. Should throw an error. (This seems like a good reason to have separate Condition subclasses for single-key conditions and sympy expressions: we can delegate the logic to them).